### PR TITLE
added full solution

### PR DESCRIPTION
**Pull request by Rudiachenko**

### CODE REVIEW

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        }
+        insertDataToOrdersProducts(order);
+        return order;
+    }
+
+    @Override
+    public Optional<Order> getById(Long id) {
+        Order order = new Order();
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "SELECT * FROM orders WHERE order_id = ?";
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]: what about deleted orders?

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                order = createOrderFromResultSet(resultSet);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Get order with id "
+                    + id + " is failed", e);
+        }
+        order.setProducts(extractProductsForOrder(order.getId()));
+        return Optional.of(order);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java:58]: you should return optional empty if order was not found

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    @Override
+    public List<Order> getAll() {
+        List<Order> orders = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "SELECT * FROM orders";
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]: check if order was deleted

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+            PreparedStatement statement = connection.prepareStatement(query);
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                Order order = createOrderFromResultSet(resultSet);
+                orders.add(order);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Get orders failed", e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]:
```suggestion
            throw new DataProcessingException("Can't get all orders", e);
```

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        return order;
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "UPDATE orders SET deleted = TRUE WHERE order_id = ?";
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, id);
+            deleteProductsFromOrder(id);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]: don't delete products from order, just mark it deleted

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+            String query = "INSERT INTO orders_products (product_id, order_id) VALUES (?, ?)";
+            PreparedStatement statement = connection.prepareStatement(query);
+            for (Product product : order.getProducts()) {
+                statement.setLong(1, product.getId());
+                statement.setLong(2, order.getId());
+                statement.executeUpdate();
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't insert order with id "
+                    + order.getId() + " is failed", e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]:
```suggestion
                    + order.getId(), e);
```

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+            statement.setLong(1, shoppingCart.getUserId());
+            statement.executeUpdate();
+            ResultSet generatedKeys = statement.getGeneratedKeys();
+            while (generatedKeys.next()) {
+                long shoppingCartId = generatedKeys.getLong(1);
+                shoppingCart.setId(shoppingCartId);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("It is impossible to create "
+                    + "a shopping cart for user id: "
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java]:
```suggestion
                    + "a shopping cart for user with id: "
```

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                shoppingCart = createShoppingCartFromResultSet(resultSet);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Get shopping cart with id "
+                    + id + " is failed", e);
+        }
+        shoppingCart.setProducts(extractProductsForShoppingCart(shoppingCart.getId()));
+        return Optional.of(shoppingCart);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java:59]: You should return optional empty when cart was not found

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                shoppingCart = createShoppingCartFromResultSet(resultSet);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Get shopping cart of user with id "
+                    + userId + " is failed", e);
+        }
+        shoppingCart.setProducts(extractProductsForShoppingCart(shoppingCart.getUserId()));
+        return Optional.of(shoppingCart);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java:80]: same

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    }
+
+    @Override
+    public List<ShoppingCart> getAll() {
+        List<ShoppingCart> shoppingCarts = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "SELECT * FROM shopping_carts";
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java]: Add check for deleted records

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+        }
+        shoppingCart.setProducts(extractProductsForShoppingCart(shoppingCart.getId()));
+        return Optional.of(shoppingCart);
+    }
+
+    @Override
+    public Optional<ShoppingCart> getByUserId(Long userId) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "SELECT * FROM shopping_carts WHERE user_id = ?";
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java]: Add check for deleted records

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+        }
+        insertDataToShoppingCartsProducts(shoppingCart);
+        return shoppingCart;
+    }
+
+    @Override
+    public Optional<ShoppingCart> getById(Long id) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "SELECT * FROM shopping_carts WHERE cart_id = ?";
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java]: Add check for deleted records

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+            String query = "INSERT INTO shopping_carts_products "
+                    + "(cart_id, product_id) VALUES (?, ?)";
+            PreparedStatement statement = connection.prepareStatement(query);
+            for (Product product : shoppingCart.getProducts()) {
+                statement.setLong(1, shoppingCart.getId());
+                statement.setLong(2, product.getId());
+                statement.executeUpdate();
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get shopping cart with id "
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java]: fix exception message

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+            while (resultSet.next()) {
+                long productId = resultSet.getLong("product_id");
+                String name = resultSet.getString("name");
+                long price = resultSet.getLong("price");
+                Product product = new Product(productId, name, price);
+                products.add(product);
+            }
+            return products;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Get shopping cart with id : "
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java]: same

File: src/main/java/com/internet/shop/dao/jdbc/UsersDaoJdbcImpl.java [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+            ResultSet generatedKeys = statement.getGeneratedKeys();
+            while (generatedKeys.next()) {
+                long userId = generatedKeys.getLong(1);
+                user.setId(userId);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("It is impossible to create a user: "
+                    + user, e);
+        }
+        return insertRoleToUsersRolesTable(user);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/UsersDaoJdbcImpl.java]: does user have ids for his roles, if yes, when did you set them ?

File: src/main/java/com/internet/shop/dao/jdbc/UsersDaoJdbcImpl.java [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                user = createUserFromResultSet(resultSet);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Get user with id "
+                    + id + " is failed", e);
+        }
+        return Optional.of(user);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/UsersDaoJdbcImpl.java]: You should return optional empty when user was not found

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+    @Override
+    public Optional<Order> getById(Long id) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "SELECT * FROM orders WHERE order_id = ? AND deleted = FALSE";
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                Order order = createOrderFromResultSet(resultSet);
+                order.setProducts(extractProductsForOrder(order.getId()));
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]: let's not open connection inside connection and close the previous statement before opening the new one

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+    @Override
+    public Optional<ShoppingCart> getById(Long id) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "SELECT * FROM shopping_carts WHERE cart_id = ? AND deleted = FALSE";
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                ShoppingCart shoppingCart = createShoppingCartFromResultSet(resultSet);
+                shoppingCart.setProducts(extractProductsForShoppingCart(shoppingCart.getId()));
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java]: same here/ You can move it after catch block or pass connection object to method

File: src/main/java/com/internet/shop/dao/jdbc/UsersDaoJdbcImpl.java [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+
+    @Override
+    public Optional<User> getById(Long id) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "SELECT * FROM users WHERE user_id = ? AND deleted = FALSE;";
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                User user = createUserFromResultSet(resultSet);
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/dao/jdbc/UsersDaoJdbcImpl.java]: you are opening a new connection in setRoles method which is inside createUserFromResultSet

Comment (dmytro-hryhoruk) [reply src/main/java/com/internet/shop/dao/jdbc/UsersDaoJdbcImpl.java]: you can divide  the logic of createUserFromResultSet and setRoles and call them both here(after catch block)

**Review by dmytro-hryhoruk — APPROVED** [review]

**Pull request merged**